snapcraft/plugins/rust.py

```diff
@@ -193,21 +194,74 @@ def _get_target(self) -> str:
             )
         return rust_target.format("unknown-linux", "gnu")
 
+    def _project_uses_workspace(self) -> bool:
+        path = Path(self.builddir, "Cargo.toml")
+        if not path.is_file():
+            return False
+
+        config = open(path).read()
+        return "workspace" in toml.loads(config)
+
+    def _install_workspace_artifacts(self) -> None:
+        """Install workspace artifacts."""
+        # Find artifacts in release directory.
+        release_dir = Path(self.builddir, "target", "release")
+
+        # Install binaries to bin/.
+        bins_dir = Path(self.installdir, "bin")
+        bins_dir.mkdir(parents=True, exist_ok=True)
+
+        # Install shared objects to usr/lib/<arch-triplet>.
+        # TODO: Dynamic library support needs to be properly added.
+        # Although weinstall libraries if we find them, they are most
+        # likely going to be missing dependencies, e.g.:
+        # /home/ubuntu/.cargo/toolchains/stable-x86_64-unknown-linux-gnu/lib/libstd-fae576517123aa4e.so
+        libs_dir = Path(self.installdir, "usr", "lib", self.project.arch_triplet)
+        libs_dir.mkdir(parents=True, exist_ok=True)
+
+        # Cargo build marks binaries and shared objects executable...
+        # Search target directory to get these and install them to the
+        # correct location.
+        for path in release_dir.iterdir():
+            if not os.path.isfile(path):
+                continue
+            if not os.access(path, os.X_OK):
+                continue
+
+            # File is executable, now to determine if bin or lib...
+            if path.name.endswith(".so"):
+                file_utils.link_or_copy(path.as_posix(), libs_dir.as_posix())
+            else:
+                file_utils.link_or_copy(path.as_posix(), bins_dir.as_posix())
+
     def build(self):
         super().build()
 
-        # Write a minimal config.
-        self._write_cargo_config()
+        uses_workspaces = self._project_uses_workspace()
+
+        if uses_workspaces:
+            # This is a bit ugly because `cargo install` does not yet support
```

Comment:
seems some of the discrepancies we have when using go.mod can also be seen here :man_shrugging:

Reply:
Ah yes, indeed! :)